### PR TITLE
AEA-1289 task manager now uses thread pool. fixes for skills to use task.

### LINKED ISSUE #2222: Expose enqueue_task, get_task_result in skills
**Is your feature request related to a problem? Please describe.**
Currently, the skill has no access to the methods enqueue_task and get_task_result.

**Describe the solution you'd like**
Make them accessible in the skill context. Then use it in the gym skill and ml skill.

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

task manager now uses thread pool cause some issues with pickling
fixes for skills to use task: gym and ml_train skills

## Fixes

#2222

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/fetchai/skills/ml_train/tasks.py [hunk 1]
```diff
@@ -35,22 +36,38 @@ def __init__(
...
+                tf.keras.layers.Dense(128, activation="relu"),
+                tf.keras.layers.Dense(10, activation="softmax"),
+            ]
+        )
+        model.compile(
+            optimizer="adam",
+            loss="sparse_categorical_crossentropy",
+            metrics=["accuracy"],
+        )
+        return model
```
Comment (DavidMinarsch) [inline packages/fetchai/skills/ml_train/tasks.py:69]: Where was this done previously?

**Review by DavidMinarsch — COMMENTED** [review]
Cannot see at the moment where the task results are actually processed ...

File: aea/skills/tasks.py [hunk 2]
```diff
@@ -109,20 +110,22 @@ def init_worker() -> None:
...
-    Disable the SIGINT handler.
+    Disable the SIGINT handler of process pool is using.
     Related to a well-known bug: https://bugs.python.org/issue8296
 
     :return: None
     """
-    signal.signal(signal.SIGINT, signal.SIG_IGN)
+    if Pool.__class__.__name__ == "Pool":  # pragma: nocover
+        # Process worker
+        signal.signal(signal.SIGINT, signal.SIG_IGN)
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/skills/tasks.py:120]: it's present in init_worker doc string

File: packages/fetchai/skills/gym/handlers.py [hunk 3]
```diff
@@ -45,13 +44,13 @@ def __init__(self, **kwargs: Any):
...
         self.task = GymTask(self.context, nb_steps)
+        self._task_id: Optional[int] = None
 
     def setup(self) -> None:
         """Set up the handler."""
         self.context.logger.info("Gym handler: setup method called.")
         # launch the task
-        self.task.setup()
-        self.task.execute()
+        self._task_id = self.context.task_manager.enqueue_task(self.task)
```
[the remark this answers is not in the document]
Comment (solarw) [reply packages/fetchai/skills/gym/handlers.py:53]: in handler teardown called task teardown.
not sure this part is correct, by i'm out of the gym skill logic
task should be quite short, not permanent.
so probably should be rewrited

File: packages/fetchai/skills/ml_train/tasks.py (hunk 1 above)
Comment (solarw) [reply packages/fetchai/skills/ml_train/tasks.py:69]: in the model, so model spawned a thread and do the calculations on update method.

File: packages/fetchai/skills/ml_train/ml_model.py [hunk 4]
```diff
@@ -135,5 +133,3 @@ def teardown(self) -> None:
 
         :return: None
         """
-        self.data_queue.put(None)
```
[the remark this answers is not in the document]
Comment (solarw) [reply packages/fetchai/skills/ml_train/ml_model.py:138]: no, it was used as task queue

**Review by DavidMinarsch — APPROVED** [review]
LGTM

**Pull request merged**